File: server/common/protocol.py
import struct


def send_all(sock, data: bytes):
    """Send all bytes to socket, avoiding short-write."""
    while data:
        sent = sock.send(data)
        data = data[sent:]


def recv_all(sock, n: int) -> bytes:
    """Receive exactly n bytes from socket, avoiding short-read."""
    buf = b""
    while len(buf) < n:
        chunk = sock.recv(n - len(buf))
        if not chunk:
            raise ConnectionError("Connection closed while receiving data")
        buf += chunk
    return buf


def send_message(sock, payload: bytes):
    """Send a length-prefixed message: 4 bytes big-endian length + payload."""
    length = struct.pack('>I', len(payload))
    send_all(sock, length)
    send_all(sock, payload)


def recv_message(sock) -> bytes:
    """Receive a length-prefixed message: 4 bytes big-endian length + payload."""
    header = recv_all(sock, 4)
    length = struct.unpack('>I', header)[0]
    return recv_all(sock, length)

File: client/common/protocol.go
package common

import (
	"encoding/binary"
	"fmt"
	"net"
)

func SendAll(conn net.Conn, data []byte) error {
	for len(data) > 0 {
		n, err := conn.Write(data)
		if err != nil {
			return err
		}
		data = data[n:]
	}
	return nil
}

func RecvAll(conn net.Conn, n int) ([]byte, error) {
	buf := make([]byte, n)
	offset := 0
	for offset < n {
		read, err := conn.Read(buf[offset:])
		if err != nil {
			return nil, err
		}
		offset += read
	}
	return buf, nil
}

func SendMessage(conn net.Conn, payload []byte) error {
	header := make([]byte, 4)
	binary.BigEndian.PutUint32(header, uint32(len(payload)))
	if err := SendAll(conn, header); err != nil {
		return fmt.Errorf("failed to send header: %w", err)
	}
	if err := SendAll(conn, payload); err != nil {
		return fmt.Errorf("failed to send payload: %w", err)
	}
	return nil
}

func RecvMessage(conn net.Conn) ([]byte, error) {
	header, err := RecvAll(conn, 4)
	if err != nil {
		return nil, fmt.Errorf("failed to read header: %w", err)
	}
	length := binary.BigEndian.Uint32(header)
	payload, err := RecvAll(conn, int(length))
	if err != nil {
		return nil, fmt.Errorf("failed to read payload: %w", err)
	}
	return payload, nil
}
